python/hpsmc/batch.py

```diff
@@ -273,45 +327,119 @@ def submit_cmd(self, name, job_params):
         return batch_id
 
 
-class Slurm(Batch):
+class Slurm(BatchSystem):
     """! Submit Slurm batch jobs."""
 
     def __init__(self):
-        os.environ["LSB_JOB_REPORT_MAIL"] = "N"
+
+        super().__init__()
+
+        self.parser.add_argument("-S", "--sh-dir", nargs='?', help="Directory to hold generated shell scripts for Slurm", default=str(Path(os.getcwd(), 'sh')))
 
     def parse_args(self, args):
-        Batch.parse_args(self, args)
-        if self.email:
-            os.environ["LSB_JOB_REPORT_MAIL"] = "Y"
 
-    def build_cmd(self, name, job_params):
-        log_file = os.path.abspath(os.path.join(self.log_dir, 'job.%s.log' % str(name)))
+        cl = super().parse_args(args)
+
+        # Set Slurm scripts dir
+        self.sh_dir = os.path.abspath(cl.sh_dir)
+        logger.info('Slurm sh dir: {}'.format(self.sh_dir))
+        if not os.path.exists(self.sh_dir):
+            os.makedirs(self.sh_dir)
+            logger.info('Created Slurm sh dir: {}'.format(self.sh_dir))
+
+    def default_rundir(self, job_id=None):
+        """!
+        Override the basic implementation for getting the default run directory.
+        """
+        if self.site == 'slac':
+            run_dir = '$LSCRATCH'
+        elif self.site == 'jlab':
+            run_dir = '/scratch/slurm/$SLURM_JOBID'
+        else:
+            run_dir = os.getcwd() + "/scratch/$SLURM_JOBID"
+        return run_dir
 
+    def _default_queue(self):
         queue = self.queue
         if queue is None:
-            queue = 'hps'
-
-        cmd = ['sbatch',
-               '--time=%s' % (str(self.job_length) + ':00:00'),
-               '--partition=%s' % queue,
-               '--mem=%sM' % self.memory,
-               '--job-name=hps%i' % (job_params['job_id']),
-               '--output=%s.out' % log_file]
-        sh_filename = self.sh_dir + '/job.%i.sh' % job_params['job_id']
-        sh_file = open(sh_filename, 'w')
-        sh_file.write('#!/usr/bin/scl enable devtoolset-8 -- /bin/bash\n')
-        sh_file.write('source ' + self.env + '\n')
-        sh_file.write('mkdir -p %s/%i\n' % (self.run_dir, job_params['job_id']))
-        sh_file.write('cd %s/%i\n' % (self.run_dir, job_params['job_id']))
-        sh_file.write(' '.join(Batch.build_cmd(self, name, job_params, set_job_dir=True)) + '\n')
-        sh_file.close()
+            if self.site == 'slac':
+                queue = 'shared'
+            elif self.site == 'jlab':
+                queue = 'ifarm'
+            else:
+                raise Exception('No queue name was provided.')
+        return queue
+
+    def _sbatch(self, job_id):
+        log_file = self._logfile(job_id)
+        sbatch_cmd = ['sbatch',
+                      '--time=%s' % (str(self.job_length) + ':00:00'),
+                      '--partition=%s' % self._default_queue(),
+                      '--mem=%sM' % self.memory,
+                      '--job-name=%s_%i' % (self.script_name, job_id),
```

Comment:
since the script names can be long, would like to have job_id first in the job name. Please change order here.

Reply:
If you have multiple workflows going on the batch system, listing the job IDs first is going to be confusing. So wouldn't it make sense to prepend with the script name? I agree they could be long and truncated in certain command line tools. Maybe this prepend is something that should be settable?

Reply:
you can usually tell which set of jobs are which by the timing, and the information not as easily obtainable is specifically the job_id

Reply:
> you can usually tell which set of jobs are which by the timing, and the information not as easily obtainable is specifically the job_id

I swapped this to `1234_slic` instead.